Humera - Taking over for Summit - Fix personal max record badge

skaush21 opened the pull request:
# Description

<img width="583" alt="Screen Shot 2025-01-02 at 3 45 53 PM" src="https://github.com/user-attachments/assets/0fa9edf6-244e-4672-903a-5533ca41357f" />

Fixes:
1. Based on the current design documentation and schema, updated the Data Source to merge Time Entries and savedTangibleHrs to get an accurate reading on the personal best max hour.
2. Modified a date format that fixed the earned dates that were not being added. 


## Related PRS (if any):
Related backend PR: [PR #947 ](https://github.com/OneCommunityGlobal/HGNRest/pull/947)
To test this PR use frontend development.

## Main changes explained:
- Fixed formatting issue, that was causing earnedDates to be inaccurate.
- optimized code, improved clarity, and reduced overhead calls.
-  added a merge function to merge `Time Entries` and `user.savedTangibleHrs`
- cleaned up the code from debugging and uncommented the functioning code.
- Fix new users not being assigned the badge, refactored code and improved added comments for readability
- Fix badgeOfType was being assigned the badgeCollection unnecessarily in the loop that checks for duplicate badges
- removed redundant promise query's

## How to test:
1. check into the current branch
4. do `npm install` and `...` to run this PR locally
5. Clear site data/cache
6. Test 1:
   a.  log in as Owner user
   b. go to dashboard→ Add Time via `Add intangible Time Entry` 
        Note: when logging hours, mark inTangible time as Tangible 
   c. Use Postman to initiate a request.
   d. Navigate to dashboard -> view Profile -> Featured Badges
   e. Verify that the New Personal Max Record Badge icon contains the current 
       personal max record
    f. verify that the badge is assigned in the badgeCollection.
Note: earnedDate will showcase the current date (Sunday) and not the previous dates, this is because `user.savedTangibleHrs` does not have any dates associated with its hours, whereas `Time Entries` does. Due to this, previous dates are not accounted for. 
Walkthrough:

https://github.com/user-attachments/assets/4292fefd-d309-4290-bae5-3dd87d374028

7. Test 2 - MongoDB:
   a.  On MongoDB navigate to userProfile -> savedTangibleHrs. 
   b. Add hours under savedTangibleHrs field.
       Note: The personal best max hour takes the max hour overall weeks. Any hour added must be higher than any 
       previous week. Also, you will need to update lastweektangiblehrs manually in order to test this (should be the 
       previous hour)
   c. Use Postman to send requests and verify in the frontend that the personal best max hour has been updated with the 
       new hour.
   d. Additionally verify that the badge appears in the badgeCollection. 
       1). Navigate to this via userProfiles -> badgeCollection -> object
       2). Copy the badgeId and in a separate tab search for badges. Paste the  badgeId in between the quotation marks in 
             the following. { _id: ObjectId("")}.
 Walkthrough: 


https://github.com/user-attachments/assets/1ae7997e-b9ae-4431-a01d-7d00ebdc8502


## Screenshots or videos of changes:

Modify the backend with the following to test this PR:

<img width="539" alt="327992191-a51b8597-35f7-4f28-a7c7-304539c3538c" src="https://github.com/user-attachments/assets/3c70fdb3-fc94-4254-a917-f2c2c35d9163" />

<img width="413" alt="327982279-4df5dd7a-0773-420a-b741-169966447fd6" src="https://github.com/user-attachments/assets/dccb4b64-f3da-48bb-ac76-eb65435c0dfe" />

<img width="397" alt="327981826-24ca6538-6e9f-4af9-ba93-9314e454b782" src="https://github.com/user-attachments/assets/ff2e943c-cf49-4c1e-a076-d47a65fc8a82" />

<img width="455" alt="327981812-431822cc-dd15-467a-8bc1-664a9c6a4bae" src="https://github.com/user-attachments/assets/2fd5625f-31cd-4b53-9b95-fa74b2ec6cd5" />

<img width="487" alt="327981804-c3aa9a18-e29f-4e00-9c93-d1050d7f5aa7" src="https://github.com/user-attachments/assets/0be5a81c-362d-4815-bc1c-8ce09eabac54" />

Verify

<img width="332" alt="Screen Shot 2025-01-02 at 5 23 02 PM" src="https://github.com/user-attachments/assets/d39ec7ca-9622-4a3c-b5dc-8d5f62d51d8d" />

<img width="539" alt="Screen Shot 2025-01-02 at 5 23 15 PM" src="https://github.com/user-attachments/assets/a4031d85-4acb-4a22-9ca7-e23b10ee13cc" />

<img width="725" alt="Screen Shot 2025-01-02 at 5 24 52 PM" src="https://github.com/user-attachments/assets/857d1dbf-f828-4746-8652-8881ecc3a6ae" />



## Note:
1. Will need to use Postman to POST and GET requests and MongoDB to update savedTangibleHrs and verify that the badge collection has the badge.

2. If for some reason you are unable to pull the correct code here is a link to the [branch](https://github.com/OneCommunityGlobal/HGNRest/tree/Fix_Personal_Max_Record_Badge).

skaush21 committed:
fixed formatting issue, that was causing earnedDates to be innacurate.

skaush21 committed:
optimized code, improved clarity, reduced overhead call.

skaush21 committed:
added a temp merge for the two tangible hour data sets

skaush21 committed:
including dates for time entries and populating savedTangibleHrs with current weeks hours.

skaush21 committed:
removed previous date implementation, cleaned up code.

skaush21 committed:
uncommented code, modified a comment

skaush21 committed:
modified comment

kobakvantrishvili reviewed (dismissed):
It's unclear to me how to test this PR. I don't know where exactly to modify the backend code and I'm also not sure what HTTP request I should send out. More clarity or a video demonstration would be good. I have no complaints with the actual code that's modified as it makes sense and is readable.

jeyanthi-sm commented:
The Personal Max Record Badge is not awarded to my test email address after following those steps. I have attached the exact steps and the resulting output.
![PR-1193](https://github.com/user-attachments/assets/b1a06ce0-ffc7-4775-92b6-cfb02e2676bd)

jeyanthi-sm reviewed (dismissed):
The Personal Max Record Badge is not awarded to my test email address after following those steps. I have attached the exact steps and the resulting output.
![PR-1193](https://github.com/user-attachments/assets/b1a06ce0-ffc7-4775-92b6-cfb02e2676bd)

If the user has no badges or joined recently, then there won’t be any badges, and this loop won’t execute at all:


for (let i = 0; i < badgeCollection.length; i += 1) {
  if (badgeCollection[i].badge?.type === 'Personal Max') {
As a result, the record for 'Personal Max' won’t be created.

The same situation applies to a new user with a default lastWeekTangibleHrs. In that case, this condition will fail:

user.lastWeekTangibleHrs &&
user.lastWeekTangibleHrs >= user.personalBestMaxHrs

Additionally, if badgeOfType is undefined, this line will result in an error:
badgeOfType.earnedDate.unshift(moment().format('MMM-DD-YYYY'));

saigirishpabbathi reviewed (dismissed):
Reviewed the PR but the personal max is not working as mentioned in the description. Also, instead of naming in a generic way please use appropriate method name like `mergeHours` or something feasible to you instead of mergeArrays.
Also do not add comments in the code.

https://github.com/user-attachments/assets/7103534f-1239-4a27-8169-1b45d9954578

skaush21 committed:
Fix new users not being assigned badge, optimized code and improved readability.

skaush21 committed:
fixed duplicatebadge check and further optimized code

skaush21 committed:
resolving merge conflict

skaush21 committed:
fixed count and earnedDate not updating

skaush21 committed:
removed changes made for debugging purposes

skaush21 commented:
> It's unclear to me how to test this PR. I don't know where exactly to modify the backend code and I'm also not sure what HTTP request I should send out. More clarity or a video demonstration would be good. I have no complaints with the actual code that's modified as it makes sense and is readable.

I've updated the PRs how to test guide and added walkthroughs. Please re-review this, thanks!

skaush21 commented:
> The Personal Max Record Badge is not awarded to my test email address after following those steps. I have attached the exact steps and the resulting output. ![PR-1193](https://private-user-images.githubusercontent.com/111299753/404196097-b1a06ce0-ffc7-4775-92b6-cfb02e2676bd.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5MDQ0NDYsIm5iZiI6MTc0MDkwNDE0NiwicGF0aCI6Ii8xMTEyOTk3NTMvNDA0MTk2MDk3LWIxYTA2Y2UwLWZmYzctNDc3NS05MmI2LWNmYjAyZTI2NzZiZC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwMlQwODI5MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYTdiMTM1MjIwOWI1ZDk1NDcyMDdjZGMwOGFjODMyYzQyZDkzZDdkMTlhY2I3MDFlNzhkNDU2OTNmNzUzMWNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NYmWUOq80oYjuVBZ1jVC_oFTVdP4Dc2dcq0KjXO-5Pc)
> 
> If the user has no badges or joined recently, then there won’t be any badges, and this loop won’t execute at all:
> 
> for (let i = 0; i < badgeCollection.length; i += 1) { if (badgeCollection[i].badge?.type === 'Personal Max') { As a result, the record for 'Personal Max' won’t be created.
> 
> The same situation applies to a new user with a default lastWeekTangibleHrs. In that case, this condition will fail:
> 
> user.lastWeekTangibleHrs && user.lastWeekTangibleHrs >= user.personalBestMaxHrs
> 
> Additionally, if badgeOfType is undefined, this line will result in an error: badgeOfType.earnedDate.unshift(moment().format('MMM-DD-YYYY'));

Hi, I've updated the code to include the changes you mentioned, new users should now be assigned a badge, I also updated the test guide which should help with testing. I also removed the unshift as it was not necessary.

skaush21 committed:
removed some comments

skaush21 commented:
> Reviewed the PR but the personal max is not working as mentioned in the description. Also, instead of naming in a generic way please use appropriate method name like `mergeHours` or something feasible to you instead of mergeArrays. Also do not add comments in the code.
> 
>  Screen.Recording.2025-01-25.at.6.16.02.PM.mov

Hi, Thank you for the review. I've fixed the issue and I made changed the name. As for the comments, I'm not sure which comments you were referring to however, I made some adjustments and removed any unnecessary comments.

humera314 commented:
Whenever I call Postman API this is the message I'm receiving 
![image](https://github.com/user-attachments/assets/196e367e-f05d-4b53-94d6-500037024d96)

skaush21 committed:
add comment

skaush21 committed:
removed redundant comment

skaush21 commented:
> Whenever I call Postman API this is the message I'm receiving ![image](https://private-user-images.githubusercontent.com/69070883/420517380-196e367e-f05d-4b53-94d6-500037024d96.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEzOTMzNjAsIm5iZiI6MTc0MTM5MzA2MCwicGF0aCI6Ii82OTA3MDg4My80MjA1MTczODAtMTk2ZTM2N2UtZjA1ZC00YjUzLTk0ZDYtNTAwMDM3MDI0ZDk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA4VDAwMTc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM5N2Q4OWQ0YzIxYWI0MDA3OWJlOTJjZjgzZjk3YWUzYTgxMDA4ZjhmMDA1NWJjOGNkNjIxMjEzYmRmZGJlYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Xf7dzbcrXq-aEkJWEU8vU3qT_d2p11PWZJZKXoDzcIk)

I reran the code and did not face this issue. I pushed the code again to see if that fixed the issue. Please pull the latest commit if possible. If that still doesn't work, here's the [branch](https://github.com/OneCommunityGlobal/HGNRest/tree/Fix_Personal_Max_Record_Badge) that this PR is linked to.

jeyanthi-sm reviewed (changes requested):
I tested the PR again but was not awarded the Personal Max badge. Here are the screenshots:
![third](https://github.com/user-attachments/assets/1be9deda-49de-4e94-ad5f-f39a6a648023)
![second](https://github.com/user-attachments/assets/ed0d0063-9ed7-4900-8679-e8c3b73a9307)
![first](https://github.com/user-attachments/assets/0c1be12d-4020-4282-a6c9-02af8566ea7b)
![TimeEntries](https://github.com/user-attachments/assets/f9b5c028-c218-4ff4-a4fe-08edaa75b60b)

CasstielP commented:
followed the instruction to add tangible time via owner/admin account, no badge was given after logging tangible time(31 hours) attempted 2 times with same result
<img width="999" alt="Screenshot 2025-05-16 at 10 22 09 AM" src="https://github.com/user-attachments/assets/d51634b0-46c4-4535-a13b-1cc6ccb9953e" />

Commit:
Add updated userhelper folder for Fix_Personal_Max_Record_Badge

humera314 committed:
Merge branch 'development' into Fix_Personal_Max_Record_Badge

humera314 committed:
Merge branch 'development' into Fix_Personal_Max_Record_Badge

humera314 committed:
Merge branch 'development' into Fix_Personal_Max_Record_Badge

sundarmachani reviewed (approved):
Tested this PR using both methods — adding work logs via the UI and manually editing the database documents. I was able to confirm that the Personal Best badge correctly updates its personal best hours based on the highest number of hours worked.

<img width="1725" alt="Screenshot 2025-06-13 at 3 36 09 PM" src="https://github.com/user-attachments/assets/cd86546d-2eef-4f6e-b27b-cd00c1636fd5" />  
<img width="1170" alt="Screenshot 2025-06-13 at 3 36 21 PM" src="https://github.com/user-attachments/assets/fe8a70eb-15bf-443a-af48-7c6314050af3" />
<img width="1721" alt="Screenshot 2025-06-13 at 3 38 04 PM" src="https://github.com/user-attachments/assets/9614e7ea-0ef0-4ba7-887f-f8466d197c86" />

one-community commented:
Redone with this PR: https://github.com/OneCommunityGlobal/HGNRest/pull/2116
